recipes/xz_utils/all/conanfile.py

```diff
@@ -28,159 +34,146 @@ class XZUtils(ConanFile):
-                "-install_name \\$rpath/",
-                "-install_name @rpath/",
-            )
-
     def _build_msvc(self):
         # windows\INSTALL-MSVC.txt
-        msvc_version = "vs2017" if Version(self.settings.compiler.version) >= "15" else "vs2013"
-        with tools.chdir(os.path.join(self._source_subfolder, "windows", msvc_version)):
+        if (self.settings.compiler == "Visual Studio" and Version(self.settings.compiler) >= "15") or \
+           (self.settings.compiler == "msvc" and Version(self.settings.compiler) >= "191"):
+            msvc_version = "vs2017"
+        else:
+            msvc_version = "vs2013"
+        with chdir(os.path.join(self.source_folder, "windows", msvc_version)):
             target = "liblzma_dll" if self.options.shared else "liblzma"
             msbuild = MSBuild(self)
-            msbuild.build(
-                "xz_win.sln",
-                targets=[target],
-                build_type=self._effective_msbuild_type,
-                platforms={"x86": "Win32", "x86_64": "x64"},
-                upgrade_project=Version(self.settings.compiler.version) >= "17")
-
-    def _configure_autotools(self):
-        if self._autotools:
-            return self._autotools
-        self._autotools = AutoToolsBuildEnvironment(self, win_bash=tools.os_info.is_windows)
-        args = ["--disable-doc"]
-        if self.settings.os != "Windows" and self.options.get_safe("fPIC", True):
-            args.append("--with-pic")
-        if self.options.shared:
-            args.extend(["--disable-static", "--enable-shared"])
+            msbuild.build_type = "Debug" if self.settings.build_type == "Debug" else "Release"
+            msbuild.platform = "Win32" if self.settings.arch == "x86" else msbuild.platform
+            msbuild.build("xz_win.sln", targets=[target])
+
+    def generate(self):
+        if is_msvc(self):
+            tc = MSBuildToolchain(self)
+            tc.generate()
         else:
-            args.extend(["--enable-static", "--disable-shared"])
-        if self.settings.build_type == "Debug":
-            args.append("--enable-debug")
-        self._autotools.configure(configure_dir=self._source_subfolder, args=args, build=False)
-        return self._autotools
+            tc = AutotoolsToolchain(self)
+            tc.configure_args.append("--disable-doc")
+            if self.settings.build_type == "Debug":
+                tc.configure_args.append("--enable-debug")
+            tc.generate()
+            env = VirtualBuildEnv(self)
+            env.generate()
 
     def build(self):
         self._apply_patches()
-        if self._is_msvc:
+        if is_msvc(self):
             self._build_msvc()
         else:
-            autotools = self._configure_autotools()
+            autotools = Autotools(self)
+            self.win_bash = True
+            autotools.configure()
             autotools.make()
+            self.win_bash = None
```

Comment:
@memsharded could you confirm this is the canonical way to conditionally enable win_bash in conan v2 methods (for this case it's for MinGW, msys2 is added to tool_requires)?

Reply:
I've tested locally on windows with MinGW. It works... almost:

It fails during installation because install path is not converted to unix path:

```
 /usr/bin/install -c -m 644  /c/users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/src/src/liblzma/api/lzma.h 'C:Usersspaceim.conandataxz_utils5.2.5__package3cbc13750c61092b2c900ee6caef68cf1cd320b4/include/.'
make[4]: Leaving directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src/liblzma/api'
make[3]: Leaving directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src/liblzma/api'
make[3]: Entering directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src/liblzma'
make[4]: Entering directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src/liblzma'
 /usr/bin/mkdir -p 'C:Usersspaceim.conandataxz_utils5.2.5__package3cbc13750c61092b2c900ee6caef68cf1cd320b4/lib'
 /usr/bin/mkdir -p 'C:Usersspaceim.conandataxz_utils5.2.5__package3cbc13750c61092b2c900ee6caef68cf1cd320b4/lib/pkgconfig'
 /bin/sh ../../libtool   --mode=install /usr/bin/install -c   liblzma.la 'C:Usersspaceim.conandataxz_utils5.2.5__package3cbc13750c61092b2c900ee6caef68cf1cd320b4/lib'
 /usr/bin/install -c -m 644 liblzma.pc 'C:Usersspaceim.conandataxz_utils5.2.5__package3cbc13750c61092b2c900ee6caef68cf1cd320b4/lib/pkgconfig'
Usage: /c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/libtool [OPTION]... [MODE-ARG]...
Try 'libtool --help' for more information.
libtool:   error: 'C:Usersspaceim.conandataxz_utils5.2.5__package3cbc13750c61092b2c900ee6caef68cf1cd320b4/lib' must be an absolute directory name
make[4]: *** [Makefile:876: install-libLTLIBRARIES] Error 1
make[4]: Leaving directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src/liblzma'
make[3]: *** [Makefile:1802: install-am] Error 2
make[3]: Leaving directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src/liblzma'
make[2]: *** [Makefile:1638: install-recursive] Error 1
make[2]: Leaving directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src/liblzma'
make[1]: *** [Makefile:422: install-recursive] Error 1
make[1]: Leaving directory '/c/Users/spaceim/.conan/data/xz_utils/5.2.5/_/_/build/3cbc13750c61092b2c900ee6caef68cf1cd320b4/build-release/src'
make: *** [Makefile:616: install-recursive] Error 1
```

Reply:
That's it. If I replace `autotools.install()` by `autotools.install(args=[f"DESTDIR={unix_path(self, self.package_folder)}"])`, it works (note: if I call conan `create -kb` it fails with `ConanException: The config 'tools.microsoft.bash:subsystem' is needed to run commands in a Windows subsystem`, funny since this config is injected by `msys2` recipe)

Reply:
Here we go: https://github.com/conan-io/conan/issues/12154 & https://github.com/conan-io/conan/issues/12153

Reply:
@SpaceIm Thank you very much for checking it. The new MSBuild is still limited in terms of features, compared to the legacy.